### PR TITLE
refactor: Query builders, drivers, entity/column/relation metadata, decorators

### PR DESCRIPTION
**Pull request by nebkat**
# Description of change

Major refactor of query builders and other parts of code, big improvements in readability and lots of duplicate code removed.

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] `npm run lint` passes with this change
- [x] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #0000`
- [x] There are new or updated unit tests validating the change
- [x] Documentation has been updated to reflect this change
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: src/error/LimitOnDeleteNotSupportedError.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+/**
+ * Thrown when user tries to build an DELETE query with LIMIT but the database does not support it.
+*/
+export class LimitOnDeleteNotSupportedError extends Error {
```
Comment (nebkat) [inline src/error/LimitOnDeleteNotSupportedError.ts:4]: Not currently used, can we merge the errors for delete and update?

File: src/query-builder/builder/DeleteQueryBuilder.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    /**
+     * Executes sql generated by query builder and returns raw database results.
+     */
+    protected async executeInsideTransaction(queryRunner: QueryRunner): Promise<DeleteResult> {
+        const [sql, parameters] = this.getQueryAndParameters();
+        const deleteResult = new DeleteResult();
+        const result = await queryRunner.query(sql, parameters);
+
+        if (this.connection.driver instanceof MysqlDriver || this.connection.driver instanceof AuroraDataApiDriver) {
```
Comment (nebkat) [inline src/query-builder/builder/DeleteQueryBuilder.ts:64]: Can this be moved to drivers or `QueryRunner`? Each driver should know its expected return formats.

File: src/query-builder/builder/ModificationQueryBuilder.ts [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+     * Creates "LIMIT" parts of SQL query.
+     */
+    protected createLimitExpression(): string {
+        let limit: number|undefined = this.expressionMap.limit;
+
+        if (limit) {
+            if (this.connection.driver instanceof MysqlDriver || this.connection.driver instanceof AuroraDataApiDriver) {
+                return " LIMIT " + limit;
+            } else {
+                throw new LimitOnUpdateNotSupportedError();
```
Comment (nebkat) [inline src/query-builder/builder/ModificationQueryBuilder.ts:303]: `LimitNotSupportedError`? https://github.com/typeorm/typeorm/pull/7058/files#r521739540

File: src/query-builder/builder/RelationQueryBuilder.ts [hunk 4]
```diff
@@ -1,13 +1,15 @@
...
+import { QueryRunner } from '../..';
+import { BroadcasterResult } from '../../subscriber/BroadcasterResult';
 
 /**
  * Allows to work with entity relations and perform specific operations with those relations.
  *
  * todo: add transactions everywhere
  */
-export class RelationQueryBuilder<Entity> extends QueryBuilder<Entity> {
+export class RelationQueryBuilder<Entity> extends QueryBuilder<Entity, any> {
```
Comment (nebkat) [inline src/query-builder/builder/RelationQueryBuilder.ts:12]: Relation builder really doesn't fit nicely with the rest of them...

File: src/query-builder/builder/SelectQueryBuilder.ts [hunk 5]
```diff
@@ -1182,43 +1128,11 @@ export class SelectQueryBuilder<Entity> extends QueryBuilder<Entity> implements
...
-            if (this.expressionMap.useTransaction === true && queryRunner.isTransactionActive === false) {
-                await queryRunner.startTransaction();
-                transactionStartedByUs = true;
-            }
-
-            this.expressionMap.queryEntity = true;
-            const entitiesAndRaw = await this.executeEntitiesAndRawResults(queryRunner);
-            this.expressionMap.queryEntity = false;
-            const count = await this.executeCountQuery(queryRunner);
-            const results: [Entity[], number] = [entitiesAndRaw.entities, count];
```
Comment (nebkat) [inline src/query-builder/builder/SelectQueryBuilder.ts:1199]: New code could be made nicer perhaps... Counts are somewhat the exception here requiring two queries in this way.

File: src/query-builder/builder/DeleteQueryBuilder.ts (hunk 2 above)
Comment (nebkat) [reply src/query-builder/builder/DeleteQueryBuilder.ts:64]: Also with a generic solution the Insert/Update/Delete builders would become even more similar in this function.

Comment (nebkat) [reply src/query-builder/builder/DeleteQueryBuilder.ts:64]: But more generally something should really be done about the `instanceof Driver` spam...

**Comment by pleerock** [comment]
I'm not sure how to review this PR since change diff isn't clear

**Comment by nebkat** [comment]
@pleerock I've broken some of the commits up into smaller parts and added extra descriptions to the commit messages. It should now be easier to follow if you review commits one by one. Let me know if you need any further clarifications, am free to chat if necessary.

**Review by nebkat — COMMENTED** [review]
.

**Comment by nebkat** [comment]
@pleerock I've refactored this to keep up with the latest changes but you can imagine it's quite a complicated one since I have modified so many files, any chance you could take a look at it now? It is becoming rather complicated to keep up with it when working on my other PRs (expression builder and upsert). I know there it is a very large diff but on the individual commit level it should actually be quite easy to review.

There are some nice benefits especially on the driver abstraction front - the number of occurrences of `driver instanceof` goes down from ~165 to ~70 and most of the remaining ones are Mongo + the complicated Oracle/SqlServer stuff. Also generally a lot of duplicate code is removed.

I also see you have upgraded TypeScript and were working on the merge `next` into `master`, could you let me know what the plan is for the near future so I can adjust the other PRs if necessary?

**Pull request closed without merging**